notebooks/text_generator.livemd

```diff
@@ -0,0 +1,920 @@
+      [0.03448275849223137],
+      [0.10344827175140381],
+      [0.48275861144065857],
+      [0.37931033968925476],
+      [0.17241379618644714],
+      [0.24137930572032928],
+      [0.7241379022598267],
+      [0.03448275849223137],
+      [0.10344827175140381],
+      [0.20689654350280762],
+      [0.8275861740112305],
+      [0.24137930572032928],
+      [0.5517241358757019],
+      [0.7586206793785095],
+      [0.7931034564971924],
+      [0.6896551847457886],
+      [0.24137930572032928],
+      [0.7241379022598267],
+      [0.03448275849223137],
+      [0.37931033968925476],
+      ...
+    ],
+    ...
+  ]
+>
+```
+
+For our train results, We will do the same. Drop the first `sequence_length` characters and then convert them to the mapping. This time we need to do something called **one-hot encoding**.
+
+The reason we want to use one hot encoding is that in our model we don't want to only return a character as the output. We want it to return the possibility of each character for the output. This way we can decide if certain possibility is good or not or even we can decide between multiple possible outputs or even discard everything if the network is not confident enough.
+
+<!-- livebook:{"force_markdown":true} -->
+
+```elixir
+# How one hot encoding works example
+iex> one_hot_encode([
+iex>  0, 
+iex>  1, 
+iex>  2
+iex> ])
+[
+  [1, 0, 0],
+  [0, 1, 0],
+  [0, 0, 1]
+]
+```
+
+In Nx, you can achieve this encoding by using this snippet
+
+<!-- livebook:{"force_markdown":true} -->
+
+```elixir
+Nx.equal(tensor, Nx.tensor(Enum.to_list(0..last_feature)))
+```
+
+To sum it up, Here is how we generate the train results.
+
+```elixir
+train_results =
+  normalized_book_text
+  |> Enum.drop(sequence_length)
+  |> Enum.map(&Map.fetch!(char_to_int, &1))
+  |> Nx.tensor()
+  |> Nx.reshape({:auto, 1})
+  |> Nx.equal(Nx.tensor(Enum.to_list(0..(characters_count - 1))))
+```
+
+<!-- livebook:{"output":true} -->
+
+```
+#Nx.Tensor<
+  u8[155920][29]
+  EXLA.Backend<cuda:0, 0.1395165607.1902510098.152324>
+  [
+    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+    [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...],
+    ...
+  ]
+>
+```
+
+```elixir
+IO.puts("Total training data: #{Nx.shape(train_data) |> elem(0)}")
+```
+
+<!-- livebook:{"output":true} -->
+
+```
+Total training data: 155920
+```
+
+<!-- livebook:{"output":true} -->
+
+```
+:ok
+```
+
+## Defining the Model
+
+```elixir
+# As the input, We expect the sequence_length characters. Each one is a number.
+model =
+  Axon.input({nil, sequence_length, 1}, "input_chars")
+  # The Long Short Term Memory layer of our network
+  |> Axon.lstm(256)
+  # Selecting only the output from the LSTM Layer
+  |> then(fn {_, out} -> out end)
+  # Since we only want the last sequence in LSTM we will slice it and select the last one
+  |> Axon.nx(fn x -> Nx.slice_along_axis(x, sequence_length - 1, 1, axis: -2) end)
```

Comment:
I think if we do this, we no longer need flatten

```suggestion
  |> Axon.nx(fn x -> x[[0..-1//1, -1, 0..-1//1]] end)
```

Reply:
This is great. I tested it and it seems the last part is not needed since we want to select everything anyways. 
So `x[[0..-1//1, -1]]` should do the trick.

Reply:
Yup, all remaining axes are sliced in full!